chainio/interface.go

```diff
@@ -0,0 +1,51 @@
+package chainio
+
+// Blockbeat defines an interface that can be used by subsystems to retrieve
+// block data. It is sent by the BlockbeatDispatcher whenever a new block is
+// received. Once the subsystem finishes processing the block, it must signal
+// it by calling NotifyBlockProcessed.
+//
+// The blockchain is a state machine - whenever there's a state change, it's
+// manifested in a block. The blockbeat is a way to notify subsystems of this
+// state change, and to provide them with the data they need to process it. In
+// other words, subsystems must react to this state change and should consider
+// being driven by the blockbeat in their own state machines.
+type Blockbeat interface {
+	// Height returns the current block height.
+	Height() int32
+}
+
+// Consumer defines a blockbeat consumer interface. Subsystems that need block
+// info must implement it.
+type Consumer interface {
+	// TODO(yy): We should also define the start methods used by the
+	// consumers such that when implementing the interface, the consumer
+	// will always be started with a blockbeat. This cannot be enforced at
+	// the moment as we need refactor all the start methods to only take a
+	// beat.
+	//
+	// Start(beat Blockbeat) error
+
+	// Name returns a human-readable string for this subsystem.
+	Name() string
+
+	// ProcessBlock takes a blockbeat and processes it. It should not
+	// return until the subsystem has updated its state based on the block
+	// data.
+	//
+	// NOTE: The consumer must try its best to NOT return an error. If an
+	// error is returned from processing the block, it means the subsystem
+	// cannot react to onchain state changes and lnd will shutdown.
+	ProcessBlock(b Blockbeat) error
```

Comment:
don't we have a quit channel?

Reply:
hmm yes in ProcessBlock, but look at `notifyAndWait`

In case we have a timeout we then probably need to close the quit channel if we don't want a context ?

Reply:
I mean yes we know that we will shutdown eventually in case one subsytem takes too long, but I think the individual subsystems should behave in a way that we don't have lingering goroutines.

Reply:
> I think the individual subsystems should behave in a way that we don't have lingering goroutines.

I think they do? Unless a new consumer decides to implement `ProcessBlock` themselves, I don't see how we would end up with lingering goroutines.

Reply:
My understanding was the following:

We call `notifyAndWait` in the concurrent case (`notifyQueues`) => which starts now 2 goroutines, the errGroup one, and the one which starts the ProcessBlock => In case the timeout hits, the processBlock goroutine will linger around until all ProcessBlocks return and trigger the `Criticalf` log which shutsdown lnd. So when suddently we decide to not shutdown lnd in case of an error in one `ProcessBlock` function, the developer needs to make sure the goroutines are cleaned up. That's why I was proposing the context in the first place. 

But maybe this is a broader design question and should not be discussed in detail now?

Reply:
> i think in general all interface calls should take a context so that they can be cancelled by the caller.

I disagree. Putting a context in ALL interfaces means this information is leaked through every implementation, which increases the complexity of the system dramatically. Instead, we should have a central place, and preferably one place to handle the state of the context. If we wanna every subsystem to have a context, which I prefer, also tried in `HarnessTest`, then we need to think about the relationship between the system's quit chan and the context's done chan. Atm I think the context's done chan can perfectly replace the quit chan.

As for the caller to cancel the call - I don't quite follow why we need that. I mean we can certainly cancel a subscription, but that canceling behavior is a state update made to the system. Are you suggesting we should have the ability to cancel the state change made from calling the methods of a subsystem?